### PR TITLE
fix: Change IdTable idColumns to read-only and add validation on access

### PR DESCRIPTION
**Pull request by bog-walk**
#### Description

**Summary of the change**:
Adds validation of `IdTable.idColumns` on every access to ensure that a `CompositeIdTable` is not defined without any.

**Detailed description**:
- **Why**:
When testing many-to-many relations with an extra column, the user might choose to use a `CompositeIdTable` as the intermediate table, with a composite primary key including a `reference()` column for each child table. In this case, since `entityId()` would not be called on either, `idColumns` would be empty. This would cause errors the first time the `id` column is accessed.

- **What**:
This PR proposes validating the id columns and throwing an exception with suggestions to use `Column.entityId()` or call `Table.addIdColumn()`.

- **How**:
    - Make `idColumns` publicly read-only and change visibility modifier of `addIdColumn()` to protected 
    - Add test for many-to-many with extra column workaround using `CompositeIdTable` that fails without validation & manual addition of id columns
    - Validate every time `idColumns` is accessed
    - Add `CompositeID` setter for right hand side `entityID` value, as shown missing based on new test

---

#### Type of Change

Please mark the relevant options with an "X":
- [X] Bug fix
- [ ] New feature
- [ ] Documentation update

Updates/remove existing public API methods:
- [ ] Is breaking change

Affected databases:
- [X] All

#### Checklist

- [X] Unit tests are in place
- [X] The build is green (including the Detekt check)
- [X] All public methods affected by my PR has up to date API docs
- [ ] Documentation for my change is up to date

---

### CODE REVIEW

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/dao/id/IdTable.kt [hunk 1]
```diff
@@ -37,19 +37,49 @@ abstract class IdTable<T : Comparable<T>>(name: String = "") : Table(name) {
...
+    private fun <F : HashSet<Column<out Comparable<*>>>> F.validate(): F {
+        if (idColumnsValidated || inTableInitialization) return this
+
+        val primaryKeyColumns = primaryKey?.columns?.toSet().orEmpty()
+        return when {
+//            primaryKeyColumns.isEmpty() || minus(primaryKeyColumns).isNotEmpty() -> {
+//                val message = "Table definition must include a primary key with all id columns: ${joinToString { it.name }}"
+//                exposedLogger.error(message)
+//                error(message)
+//            }
```
Comment (bog-walk) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/dao/id/IdTable.kt:54]: Validation is dependent on timing in table definition (whether `entityId()` gets invoked first, or maybe primary key, so I opted to check on every access and prevent validation only if it's been confirmed.

I thought the above logic made sense, that a table with an `entityId()` column must have it as a primary key, but there are a couple examples in our tests that do not follow this rule and do not have a primary key at all. I will remove comments if we agree to merge this PR.

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/dao/id/IdTable.kt [hunk 2]
```diff
@@ -37,19 +37,49 @@ abstract class IdTable<T : Comparable<T>>(name: String = "") : Table(name) {
 
     /** All base columns that make up this [IdTable]'s identifier column. */
     val idColumns = HashSet<Column<out Comparable<*>>>()
+        get() = field.validate()
```
Comment (bog-walk) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/dao/id/IdTable.kt:40]: If anybody else has thoughts about the mutability of `idColumns`, I'm open to alternatives.

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/dao/id/IdTable.kt [hunk 3]
```diff
@@ -37,19 +37,49 @@ abstract class IdTable<T : Comparable<T>>(name: String = "") : Table(name) {
...
+//                val message = "Table definition must include a primary key with all id columns: ${joinToString { it.name }}"
+//                exposedLogger.error(message)
+//                error(message)
+//            }
+            this == primaryKeyColumns -> {
+                idColumnsValidated = true
+                this
+            }
+            else -> {
+                primaryKeyColumns.minus(this).forEach { missingColumn ->
```
Comment (obabichevjb) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/dao/id/IdTable.kt:60]: Above you told that:

> I thought the above logic made sense, that a table with an entityId() column must have it as a primary key, but there are a couple examples in our tests that do not follow this rule and do not have a primary key at all.

Which means that if the table has an entity id (or compsite id) it does not mean that table must have same primary key, sounds logical right.

But I want to make sure (I trust that you already checked it, but it does not looks obvious for me yet) that with the statement below you say, that if table has primary key, all columns of that primary key must be in the `idColumns`, is it so?

Comment (bog-walk) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/dao/id/IdTable.kt:60]: Yes, I assumed that if an `IdTable` has a primary key defined, then all columns of that primary key must also be in `idColumns`.

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/dao/id/IdTable.kt (hunk 2 above)
Comment (bog-walk) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/dao/id/IdTable.kt:40]: All `IdTable` are available within the core module. So users can access them, for convenience, even if they don't use DAO. So a user that already has been using composite primary keys in DSL could choose now to convert it to a `CompositeIdTable` to take advantage of the `id` column. So there needs to be a way to do that. If `entityid()` is moved to DAO, how will it initialize all the `IdTable` with final properties?

**Review by e5l — APPROVED** [review]

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/dao/id/IdTable.kt (hunk 3 above)
Comment (obabichevjb) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/dao/id/IdTable.kt]: > But in the original PR, you made a comment that it should not be possible to create multiple entityId() on a non-CompositeIdTable, so I don't see how the first statement is logical.

Multiple `entityId` can be used only with `CompositeIdTable`, I didn't want to say that non-`CompositeIdTable` should be able to have more that one id column

The question was about validation. Two questions are:
- must all the `idColumns` columns be in `primaryKey`?
- must all the `primaryKey` columns be in `idColumns`?

Now from your comments I see that answer to the first one is No, and to the second one is Yes.

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/dao/id/IdTable.kt (hunk 2 above)
Comment (obabichevjb) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/dao/id/IdTable.kt]: I got the point, `CompositeIdTable` as an better alternative to simple `Table` with multiple columns in primary key sounds good and similar to other `IdTables`.

**Pull request merged**